kf/kf_unix.go

```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
 	}
 	return nil
 }
+
+func AttachXDP(bpfRootProg *ebpf.Program, index int) error {
+	// Attach the program
+	_, err := link.AttachXDP(link.XDPOptions{
+		Program:   bpfRootProg,
+		Interface: index,
+	})
+
+	return err
+}
+
+// LoadTCRootProgram - Load and add tc filters
+func (b *BPF) LoadTCRootProgram(ifaceName, direction string, eBPFProgram *BPF) error {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("look up network iface %q: %s", ifaceName, err)
+	}
+
+	// verify and add attribute clsact
+	tcgo, err := tc.Open(&tc.Config{})
+	if err != nil {
+		return fmt.Errorf("could not open rtnetlink socket: %v", err)
+	}
+
+	clsactFound := false
+	// get all the qdiscs from all interfaces
+	qdiscs, err := tcgo.Qdisc().Get()
+	if err != nil {
+		return fmt.Errorf("could not get qdiscs: %v", err)
+	}
+	for _, qdisc := range qdiscs {
+		iface, err := net.InterfaceByIndex(int(qdisc.Ifindex))
+		if err != nil {
+			return fmt.Errorf("could not get interface from id %d: %v", qdisc.Ifindex, err)
+		}
+		if iface.Name == ifaceName && qdisc.Kind == "clsact" {
+			clsactFound = true
+		}
+	}
+
+	if !clsactFound {
+		qdisc := tc.Object{
+			Msg: tc.Msg{
+				Family:  unix.AF_UNSPEC,
+				Ifindex: uint32(iface.Index),
+				Handle:  core.BuildHandle(tc.HandleRoot, 0x0000),
+				Parent:  tc.HandleIngress,
+				Info:    0,
+			},
+			Attribute: tc.Attribute{
+				Kind: "clsact",
+			},
+		}
+
+		if err := tcgo.Qdisc().Add(&qdisc); err != nil {
+			log.Info().Msgf("could not assign clsact to %s: %v, its already exists", ifaceName, err)
+		}
+	}
+
+	CollectionRef, err := ebpf.LoadCollection(eBPFProgram.Program.ObjectFile)
+	if err != nil {
+		log.Error().Msgf("loading of tc root %s failed direction %s", eBPFProgram.Program.ObjectFile, direction)
+		return fmt.Errorf("%s: loading of tc root failed", eBPFProgram.Program.ObjectFile)
+	}
+
+	// storing collection reference pointer
+	b.CollectionRef = CollectionRef
+
+	var bpfRootProg *ebpf.Program
+	var bpfRootMap *ebpf.Map
+	var rootArrayMapFileName string
+
+	bpfRootProg = CollectionRef.Programs[eBPFProgram.Program.EntryFunctionName]
+	bpfRootMap = CollectionRef.Maps[strings.Split(eBPFProgram.Program.MapName, "/")[2]]
+	rootArrayMapFileName = filepath.Join(b.hostConfig.BpfMapDefaultPath, eBPFProgram.Program.MapName)
+
+	// Pinning root program
+	if err := bpfRootMap.Pin(rootArrayMapFileName); err != nil {
+		return fmt.Errorf("%s failed to pin the map", rootArrayMapFileName)
+	}
+
+	var parent uint32
+	if direction == models.IngressType {
+		parent = tc.HandleMinIngress
+	} else if direction == models.EgressType {
+		parent = tc.HandleMinEgress
+	}
+
+	progFD := uint32(bpfRootProg.FD())
+	bpfFlag := uint32(0x1)
```

Comment:
what does the number 0x1 actually mean here?  Preferably make it a symbol.  Especially just in case the value might vary by platform (Linux, Windows, whatever)